### DISCUSSION TITLE
New template with Termwind

### DISCUSSION THREAD

**Discussion by WendellAdriel** [minicli/minicli#65] [Ideas]
**[Termwind](https://github.com/nunomaduro/termwind)** is a really great package to create beautiful CLI applications and Minicli is an amazing package to create CLI applications, so why not join both?

I was thinking of creating an official template like **[minicli/application](https://github.com/minicli/application)**, something like `minicli/termwind` or `minicli/miniterm`, that will have the same structure as the minicli/application but with Termwind already installed and configured.

I'd love to hear your thoughts on this.
Especially from you both, @JustSteveKing and @erikaheidi

If you agree with this, I can start working on this tomorrow.

**Comment by JustSteveKing** [comment]
Unfortunately the aim of minicli is to be dependency free, so pulling in termwind would be a no. However, creating an "official" template that uses it could be cool!

**Reply by WendellAdriel** [reply]
Yeah, I was thinking about a template because the framework itself can't have dependencies.

**Reply by WendellAdriel** [reply]
I'm going to create a repository with this and share it with you ASAP and if you think it's a good idea we can discuss making it "official" to be under the Minicli org.

**Reply by erikaheidi** [reply]
Love the idea. As Steve said, we'll need to keep the current application template as-is but we can definitely have another application template that integrates Termwind. I am curious to try it out 😁

**Reply by WendellAdriel** [reply]
I’m going to work on it and share it here ASAP! Thanks for the feedback @JustSteveKing and @erikaheidi!!!

**Comment by WendellAdriel** [comment]
@JustSteveKing @erikaheidi here's the template:
https://github.com/WendellAdriel/miniterm

I'll be waiting for some feedback.
I'll be happy to work on any changes needed for the docs.
Also if this becomes an "official" template let me know how I can transfer that to the minicli org

**Reply by WendellAdriel** [reply]
I think that the only thing needed is to review the tests, I still need to check why they are failing and fix them

**Reply by WendellAdriel** [reply]
@JustSteveKing @erikaheidi I just pushed some new commits doing some changes in the template.
I wrapped the needed Termwind and Twig functionality into services, created a BaseController that will use these services for a better DX and fixed the issues with the test suite.

Let me know what do you think when you check it out!

**Comment by erikaheidi** [comment]
This looks really awesome @WendellAdriel ! I wonder there would be a way to overwrite the default Output / Printer to also use Termwind with some default values, so that imported commands (such as the help command) would be printed in the same style.

**Reply by WendellAdriel** [reply]
Thanks, @erikaheidi!!!
I didn't check about overwriting the default Output/Printer.
Tomorrow if I have some time I'll check and test some things and I'll let you know here!

**Reply by WendellAdriel** [reply]
@erikaheidi just pushed a commit creating a TermwindOutputHandler, please take a look when you have the chance and let me know what you think.

**Reply by erikaheidi** [reply]
I like it a lot, and the labels are super cool for logging purposes, but I think they can make it harder for migrating an existing Minicli app and commands to the Miniterm template. Take for instance the help command, we don't want to show the labels before each output line. If you can find a way to make the labels optional, I think that would be more versatile if you already have some commands implemented in "the old way". I do like the labels just to be clear, just think they shouldn't show up by default. Thanks again for working on this, it's amazing!

**Reply by WendellAdriel** [reply]
Oh true, I forgot about that.
I was already planning a little refactor on the output handler so I'm going to apply a configuration to enable/disable those as well!
Thanks for the feedback!

**Comment by JustSteveKing** [comment]
This is definitely super cool! Are there alternative options other than using twig - or is it a requirement?

**Reply by WendellAdriel** [reply]
We can use any template engine, I chose Twig because it's really simple to integrate in any PHP project and it's simple to use

**Reply by JustSteveKing** [reply]
It would be nice to see if we can reduce this even further to just PHP - see if it is possible? Or Plates (the PHP League template engine)

**Reply by WendellAdriel** [reply]
I didn't want to reduce to just PHP just for the sake of not having to create directives for loops, conditionals, etc.
I'll take a look and do some tests with Plates and I'll let you know here, thanks for the suggestions!

**Reply by erikaheidi** [reply]
Agree that Twig is super easy to integrate - I also use it for other projects. But I didn't know about Plates. Blade is a no-go for me, because it is very complex to integrate into a standalone project like this. Let us know how it goes with Plates, I'm now curious about it..

**Reply by WendellAdriel** [reply]
Yeah, I thought about Twig because it's really simple to integrate, Blade outside Laravel is more complex. As soon as I have tested Plates I'll let you know

**Reply by JustSteveKing** [reply]
Plates is an interesting one, it is a minimal wrapper around PHP itself so a good middle ground. It was more of a question than a request, tbh - and I completely agree with Blade. It is too complicated and perhaps a little too heavy for what we need.

**Reply by WendellAdriel** [reply]
I liked the suggestion tho, @JustSteveKing and I really appreciate all feedback

**Comment by erikaheidi** [comment]
This is looking good. In terms of migrating to the Minicli org, would that be ok if I create a `minicli/miniterm` repo and give @WendellAdriel admin permissions?

**Reply by WendellAdriel** [reply]
That would be ok for me if it is for you both, I don't mind of being a maintainer to that repo

**Comment by WendellAdriel** [comment]
@erikaheidi @JustSteveKing I just pushed some commits to `miniterm` with these changes:

- Created a Config folder that now has the `TermwindOutputHandler` and a new class `TermwindOutputConfig` that will be used to customize the CSS classes to be used with each style, enable and disable the labels and list which styles you want the label to be added (if they are enabled)

- Replaced `Twig` with `Plates`

**Reply by WendellAdriel** [reply]
Oh and also added a new command: `demo color` that's a version of the **[script here](https://docs.minicli.dev/en/latest/output/using-themes)** to test the themes, so you can see how each style will be printed to the terminal

**Comment by WendellAdriel** [comment]
This was now launched: https://github.com/minicli/miniterm
